Prerelease improvements

reginleif888 opened the pull request.

reginleif888 committed:
Test journal events

reginleif888 committed:
Withdrawal improvements

reginleif888 committed:
Remove minima

reginleif888 committed:
Introduce surcharge

reginleif888 committed:
Update doc comments

reginleif888 committed:
Introduce new integrity assertion

reginleif888 committed:
Take fee difference into account in consolidation

reginleif888 committed:
Rework deposit via allowance

reginleif888 committed:
Introduce the credit registry cleanup

reginleif888 committed:
Rework notifications tests

reginleif888 committed:
Test surcharge to notifications tests

reginleif888 committed:
Rework consolidation tests

reginleif888 committed:
Rework deposit via allowance tests

reginleif888 committed:
Add test for multiple consolidations trigger

reginleif888 committed:
Update allowance tests

reginleif888 committed:
Rework withdrawal tests

reginleif888 committed:
Fix allowance tests

reginleif888 committed:
Move all other tests to a separate file

reginleif888 committed:
Remove unnecessary files

reginleif888 committed:
Test surcharge and fees

reginleif888 committed:
Fix doc comment for credited

reginleif888 committed:
Update example

reginleif888 committed:
Rework README.md

reginleif888 committed:
Update doc comments

timohanke commented on README.md:
````diff
@@ -1,67 +1,153 @@
-mops install
-DFX_MOC_PATH=<path-to-moc> mops test
-```
+It is not required that the service can inspect individual deposit transactions by transaction id, memo or other means. Hence, it is not required that the underlying token ledger provides an indexer, transaction history or archive. In particular, the ICRC-3 extension is not required.
````
We should probably say somewhere that a service can support multiple tokens but has to operate a separate instance of TokenHandler for each supported token.

timohanke commented on README.md:
```diff
@@ -1,67 +1,153 @@
+
+Fees are made up of two components: the surcharge and the ledger fee.
+
+Depending on the operation, fees are calculated as follows:
+
+| Operation | Fee value |
+| --- | --- |
+| Deposit via direct transfer | surcharge + ledger_fee |
+| Deposit via an allowance | surcharge |
+| Withdrawal from credit | surcharge + ledger_fee |
+| Withdrawal from pool | ledger_fee |
```
The surcharge isn't in the code, is it? The TokenHandler allows the user to set arbitrary fees for each operation.

timohanke commented on src/IntMap.mo:
```diff
@@ -28,12 +28,26 @@ module {
+    /// Clears the map of entries with values below the provided minimum.
+    /// The method is designed to prevent spam attacks.
+    /// Returns the number of entries cleared.
+    public func clean(min : Nat) : Int {
```
Why is the type `Int`, not `Nat`?

timohanke commented on src/CreditRegistry.mo:
```diff
@@ -95,6 +95,11 @@ module {
+    /// Clears the map of entries with values below the provided minimum.
+    /// The method is designed to prevent spam attacks.
+    /// Returns the number of entries cleared.
+    public func clean(min : Nat) : Int = map.clean(min);
```
Why is the type Int, not Nat?

timohanke commented on src/lib.mo:
```diff
@@ -344,14 +344,19 @@ module {
+    /// Clears the credit registry of entries with values below the provided minimum.
+    /// The method is designed to prevent spam attacks.
+    /// Returns the number of entries cleared.
+    public func clean(min : Nat) : Int = creditRegistry.clean(min);
```
Is it clear that this is safe to do? Without looking at locks I mean.

reginleif888 replied:
In general, considering that we do pre-issue/pre-burn in notifications and withdrawals, this is not particularly safe in theory. However, our locks are in the `depositRegistry`. Wouldn't it be better to create a separate data structure with locks? Another option, probably more reliable and safer, is to make sure everything is paused before cleaning.

timohanke replied:
Let's remove `clean` functionality from this PR and introduce it later. First release of TokenHandler does not need it. Then we have more time to think about its safety.

timohanke commented on src/lib.mo:
```diff
@@ -105,30 +105,21 @@ module {
       triggerOnNotifications,
       freezeTokenHandler,
-      func(p : Principal, x : Int) { creditRegistry.issue(#user p, x) },
+      creditRegistry,
```
I would like to avoid passing creditRegistry to AccountManager. Original idea was to have AccountManager and CreditRegistry two independent modules used by TokenHandler. Now CreditRegistry becomes a submodule of AccountManager. 

Anyway, this shouldn't block this PR. We can think about and refactor this later.

timohanke commented on src/AccountManager.mo:
```diff
@@ -331,54 +250,57 @@ module {
+    func processAllowance(p : Principal, account : ICRC1.Account, amount : Nat, expectedFee : ?Nat) : async* DepositFromAllowanceResponse {
+      switch (expectedFee) {
+        case null {};
+        case (?f) if (f != fee(#allowance)) return #err(#BadFee { expected_fee = fee(#allowance) });
```
Isn't is it clearer to use `surcharge_` here in place of `fee(#allowance)` throughout the whole function? I know they are equal. But this code is the reason that they are equal. If they weren't equal then this code would be wrong.

EDIT: May be superseded by offline discussion (change of expectedFee meaning).

timohanke commented on src/AccountManager.mo:
```diff
@@ -331,54 +250,57 @@ module {
 
-      // catch BadFee
       switch (res) {
+        case (#ok txid) #ok(amount - fee(#allowance), txid);
```
This return value `amount - fee(#allowance)` is useless. The calling code is ignoring it anyway. And if the calling code needed it then it could calculate it itself from the arguments because it is equal to `amount - expectedFee` (if #ok is returned).

timohanke commented on src/AccountManager.mo:
```diff
@@ -426,100 +352,115 @@ module {
-    func withdrawRecursive(to : ICRC1.Account, amount : Nat, retry : Bool) : async* WithdrawResponse {
-      if (amount < minimum(#withdrawal)) return #err(#TooLowQuantity);
+    /// Proccesses withdrawal transfer.
+    func proccessWithdrawTransfer(p : ?Principal, to : ICRC1.Account, amount : Nat, expectedFee : ?Nat) : async* WithdrawResponse {
```
Typo: process with one c

timohanke reviewed (commented):
Left some comments

timohanke committed:
Update icrc84.did (#14)

example/mops.toml now uses icrc84#0.0.1
CI automatically uses icrc84.did from example/mops.toml

timohanke committed:
use icrc-84 repo with empty src directory (#15)

This satisfies `moc` when looking at the icrc-84 "module" installed by mops. Without the `src` directory we will see warnings in VSCode.

Also fix tag_action which accidentally triggered

reginleif888 committed:
Fix clean() return type

reginleif888 committed:
Mention multiple token support in README

reginleif888 committed:
Fix typo

reginleif888 committed:
Remove #LedgerBadFee error variant

reginleif888 committed:
Improve allowance

reginleif888 committed:
Remove clean() functionality

timohanke reviewed (approved):
LGTM
